mdpow/forcefields.py

```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
 
-    For ``solvent_type`` *water*: either provide ``None`` or "water"
-    for the specific ``model`` (and the default
-    :data:`DEFAULT_WATER_MODEL` will be selected, or a specific water
-    model such as "tip3p" or "spce" (see
-    :data:`GROMACS_WATER_MODELS`). For other "octanol" or "wetoctanol"
-    of OPLS-AA forcefield, the ``model`` is used to select a specific
-    model. For other solvents and forcefields, "model" is not required.
+    :Raises ValueError: If there is no identifier found for the combination.
 
-    :Returns: Either an identifier or ``None``
+    :Returns: An identifier
     """
+    if isinstance(forcefield, str):
+        forcefield = _get_forcefield(forcefield)
 
     if solvent_type == "water":
-        identifier = model if not model in (None, "water") else DEFAULT_WATER_MODEL
-        return identifier if identifier in GROMACS_WATER_MODELS else None
-    if model not in GROMACS_SOLVENT_MODELS[forcefield]:
-        if solvent_type in GROMACS_SOLVENT_MODELS[forcefield]:
+        identifier = (
+            forcefield.default_water_model if model in [None, "water"] else model
+        )
+
+        if identifier in forcefield.water_models:
+            return identifier
+        else:
+            raise ValueError(
+                f"{identifier} is not a valid water model for {forcefield.name}."
+            )
+
+    if model not in forcefield.solvent_models:
+        if solvent_type in forcefield.solvent_models:
             model = solvent_type
         else:
-            model = None
+            raise ValueError(
+                f"Solvent type {solvent_type} not available in {forcefield.name} solvent models."
+            )
+
     return model
 
 
-def get_solvent_model(identifier, forcefield="OPLS-AA"):
+def get_solvent_model(identifier, forcefield: Union[Forcefield, str] = OPLS_AA):
     """Return a :class:`GromacsSolventModel` corresponding to identifier *identifier*.
 
-    If identifier is "water" then the :data:`DEFAULT_WATER_MODEL` is assumed.
+    If identifier is "water" then the default water model for the :class:`Forcefield` is assumed.
+
     """
+    if isinstance(forcefield, str):
+        forcefield = _get_forcefield(forcefield)
 
     if identifier == "water":
-        identifier = DEFAULT_WATER_MODEL
+        identifier = forcefield.default_water_model
     try:
-        return GROMACS_WATER_MODELS[identifier]
+        return forcefield.water_models[identifier]
     except KeyError:
         try:
-            return GROMACS_SOLVENT_MODELS[forcefield][identifier]
+            return forcefield.solvent_models[identifier]
         except KeyError:
-            msg = "No solvent model with name {0} is available.".format(identifier)
+            msg = "No solvent model with name {0} is available for forcefield {1}.".format(
+                identifier, forcefield.name
+            )
             logger.critical(msg)
             raise ValueError(msg)
 
 
-def get_ff_paths(forcefield="OPLS-AA"):
```

Comment:
nice, I like removing whole functions — the new organization by FF makes more sense!